internal/appsec/limiter_test.go

```diff
@@ -0,0 +1,251 @@
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestLimiterUnit(t *testing.T) {
+	ticksChan := make(chan time.Time)
+	defer close(ticksChan)
+	ticker := time.Ticker{C: ticksChan}
+	defer ticker.Stop()
+	startTime := time.Now()
+
+	t.Run("no-ticks-1", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		//No ticks between the requests
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False")
+	})
+
+	t.Run("no-ticks-2", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		//No ticks between the requests
+		for i := 0; i < 10; i++ {
+			require.True(t, l.Allow(), "Call to limiter.Allow() should return True")
+		}
+	})
+
+	t.Run("10ms-ticks", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		ticksChan <- startTime.Add(10 * time.Millisecond)
+		require.True(t, l.Allow(), "Second call to limiter.Allow() after 11ms should return True")
+	})
+
+	t.Run("9ms-ticks", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		ticksChan <- startTime.Add(9 * time.Millisecond)
+		require.False(t, l.Allow(), "Second call to limiter.Allow() after 9ms should return False")
+	})
+
+	t.Run("1s-rate", func(t *testing.T) {
+		l := NewTokenTicker(1, 1)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True with 1s per token")
+		ticksChan <- startTime.Add(500 * time.Millisecond)
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False with 1s per Token")
+	})
+
+	t.Run("100-requests-burst", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		for i := 0; i < 100; i++ {
+			require.True(t, l.Allow(),
+				"Burst call %d to limiter.Allow() should return True with 100 initial tokens", i)
+			startTime = startTime.Add(50 * time.Microsecond)
+			ticksChan <- startTime
+		}
+	})
+
+	t.Run("101-requests-burst", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		for i := 0; i < 100; i++ {
+			require.True(t, l.Allow(),
+				"Burst call %d to limiter.Allow() should return True with 100 initial tokens", i)
+			startTime = startTime.Add(50 * time.Microsecond)
+			ticksChan <- startTime
+		}
+		require.False(t, l.Allow(),
+			"Burst call 101 to limiter.Allow() should return False with 100 initial tokens")
+	})
+
+	t.Run("bucket-refill-short", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+
+		for i := 0; i < 1000; i++ {
```

Comment:
Why `1000` here?

Reply:
It could be any duration really. We just want to check that the bucket doesn't overflow when refilled with the ticker waking up on short intervals. 1000 loops makes the test simulate a ticker refilling every 1ms on a 1sec total duration (since we manually tick every ms).